### PR TITLE
Fix Flowable.blockingSubscribe is unbounded and can lead to OOME

### PR DESCRIPTION
**Pull request by RomanWuattier**
Create and bound new `blockingSubscribe` overloads to `bufferSize`.
* Create new overloads with `bufferSize`
* Create a `boundedConsumer`
* Create a `BoundedSubsciber`

Close: #5988

### CODE REVIEW

File: src/main/java/io/reactivex/Flowable.java [hunk 1]
```diff
@@ -5853,6 +5853,37 @@ public final void blockingSubscribe(Consumer<? super T> onNext) {
...
+     * <dl>
+     *  <dt><b>Backpressure:</b></dt>
+     *  <dd>The operator consumes the source {@code Flowable} in an bounded manner (up to bufferSize
+     *  outstanding request amount for items).</dd>
+     *  <dt><b>Scheduler:</b></dt>
+     *  <dd>{@code blockingSubscribe} does not operate by default on a particular {@link Scheduler}.</dd>
+     * </dl>
+     * @param onNext the callback action for each source value
+     * @param bufferSize the size of the buffer
+     * @since 2.0
```
Comment (akarnokd) [inline src/main/java/io/reactivex/Flowable.java]: Please designate as `2.1.15 - experimental` here and to the other methods.

File: src/main/java/io/reactivex/Flowable.java [hunk 2]
```diff
@@ -5853,6 +5853,37 @@ public final void blockingSubscribe(Consumer<? super T> onNext) {
...
+     * @see #blockingSubscribe(Consumer, Consumer)
+     * @see #blockingSubscribe(Consumer, Consumer, Action)
+     */
+    @BackpressureSupport(BackpressureKind.FULL)
+    @SchedulerSupport(SchedulerSupport.NONE)
+    public final void blockingSubscribe(Consumer<? super T> onNext, int bufferSize) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/Flowable.java]: Please add `@Experimental` annotation to these new methods.

File: src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+public final class BoundedSubscriber<T> extends AtomicReference<Subscription>
+        implements FlowableSubscriber<T>, Subscription, Disposable, LambdaConsumerIntrospection {
+
+    private static final long serialVersionUID = -7251123623727029452L;
+    final Consumer<? super T> onNext;
+    final Consumer<? super Throwable> onError;
+    final Action onComplete;
+    final Consumer<? super Subscription> onSubscribe;
+
+    private int bufferSize;
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java]: Please no private instance fields.

File: src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    private final AtomicInteger internalBuffer;
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java]: Why is this atomic?

File: src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    }
+
+    @Override
+    public void onNext(T t) {
+        if (!isDisposed()) {
+            try {
+                if (internalBuffer.getAndIncrement() < bufferSize) {
+                    onNext.accept(t);
+                }
+                if (internalBuffer.get() == bufferSize) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java]: Why would you end the flow when the bufferSize number of elements have been received?

File: src/test/java/io/reactivex/internal/operators/flowable/FlowableBlockingTest.java [hunk 6]
```diff
@@ -121,6 +185,56 @@ public void run() throws Exception {
...
+            }
+        };
+
+        Flowable.range(1, 5)
+                .subscribeOn(Schedulers.computation())
+                .blockingSubscribe(cons, cons, action, 3);
+
+        assertEquals(Arrays.asList(1, 2, 3, 100), list);
+    }
+
```
Comment (akarnokd) [inline src/test/java/io/reactivex/internal/operators/flowable/FlowableBlockingTest.java]: Please add an unit test that transfers 1 million items.

**Review by akarnokd — DISMISSED** [review]
There are some small missing pieces and one logical error.

File: src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                RxJavaPlugins.onError(new CompositeException(t, e));
+            }
+        } else {
+            RxJavaPlugins.onError(t);
+        }
+    }
+
+    @Override
+    public void onComplete() {
+        if (get() != SubscriptionHelper.CANCELLED) {
```
Comment (davidmoten) [inline src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java:105]: if we are going to check the value of `get()` may as well do a compareAndSet loop for concurrent cancellation?

Comment (akarnokd) [reply src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java:105]: Don't. That level of precision it is not really worth it here.

File: src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java [hunk 8]
```diff
@@ -69,12 +67,14 @@ public void onSubscribe(Subscription s) {
...
-                if (internalBuffer.getAndIncrement() < bufferSize) {
-                    onNext.accept(t);
-                }
-                if (internalBuffer.get() == bufferSize) {
-                    onComplete.run();
-                    get().cancel();
+                onNext.accept(t);
+
+                int c = consumed + 1;
+                if (c == bufferSize) {
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java]: It is better to use a low-watermark than requesting only when the entire buffer has been consumed. Define a `final int limit` field, in the constructor set it as `this.limit = bufferSize - (bufferSize >> 2);`, then compare and request as `limit`.

**Review by akarnokd — COMMENTED** [review]
One algorithmic adjustment and good to go.

**Comment by akarnokd** [comment]
Also it would be great if you copied the unit tests that crash the callbacks so that the coverage on the new class gets very high.

**Comment by akarnokd** [comment]
https://codecov.io/gh/ReactiveX/RxJava/pull/6026/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9pby9yZWFjdGl2ZXgvaW50ZXJuYWwvc3Vic2NyaWJlcnMvQm91bmRlZFN1YnNjcmliZXIuamF2YQ==

**Comment by RomanWuattier** [comment]
I created a `BoundedSubscriberTest.java` to trigger `Consumer` and `Action` exceptions, but the code coverage remains unchanged. Did I miss a test?

File: src/main/java/io/reactivex/internal/operators/flowable/FlowableBlockingSubscribe.java [hunk 9]
```diff
@@ -108,4 +108,23 @@ private FlowableBlockingSubscribe() {
...
+    /**
+     * Subscribes to the source and calls the given actions on the current thread.
+     * @param o the source publisher
+     * @param onNext the callback action for each source value
+     * @param onError the callback action for an error event
+     * @param onComplete the callback action for the completion event.
+     * @param bufferSize the number of elements to prefetch from the source Publisher
+     * @param <T> the value type
+     */
+    @SuppressWarnings("ResultOfMethodCallIgnored")
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/operators/flowable/FlowableBlockingSubscribe.java]: Why is this needed. It never appears anywhere else in RxJava. Please remove.

File: src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+
+    public BoundedSubscriber(Consumer<? super T> onNext, Consumer<? super Throwable> onError,
+                            Action onComplete,
+                            Consumer<? super Subscription> onSubscribe) {
+        super();
+        this.onNext = onNext;
+        this.onError = onError;
+        this.onComplete = onComplete;
+        this.onSubscribe = onSubscribe;
+        this.bufferSize = ((Functions.BoundedConsumer) onSubscribe).getBufferSize();
```
Comment (akarnokd) [inline src/main/java/io/reactivex/internal/subscribers/BoundedSubscriber.java]: Why is this cast needed? Can't you could simply pass the bufferSize into the constructor?

File: src/test/java/io/reactivex/internal/subscribers/BoundedSubscriberTest.java [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+
+            BoundedSubscriber<Object> o = new BoundedSubscriber<Object>(new Consumer<Object>() {
+                @Override
+                public void accept(Object o) throws Exception {
+                    received.add(o);
+                }
+            }, new Consumer<Throwable>() {
+                @Override
+                public void accept(Throwable throwable) throws Exception {
+                    throw new TestException("Inner");
```
Comment (akarnokd) [inline src/test/java/io/reactivex/internal/subscribers/BoundedSubscriberTest.java:128]: Place a breakpoint here and debug the test. It should reveal why the code coverage doesn't seem to execute that code in `BoundedSubscriber`.

**Review by akarnokd — CHANGES_REQUESTED** [review]

File: src/test/java/io/reactivex/internal/subscribers/BoundedSubscriberTest.java [hunk 12]
```diff
@@ -1,11 +1,11 @@
 /**
  * Copyright (c) 2016-present, RxJava Contributors.
- *
+ * <p>
```
Comment (akarnokd) [inline src/test/java/io/reactivex/internal/subscribers/BoundedSubscriberTest.java]: Please don't change the license headers.

File: src/test/java/io/reactivex/internal/subscribers/BoundedSubscriberTest.java [hunk 13]
```diff
@@ -0,0 +1,388 @@
+/**
+ * Copyright (c) 2016-present, RxJava Contributors.
+ * <p>
```
Comment (akarnokd) [inline src/test/java/io/reactivex/internal/subscribers/BoundedSubscriberTest.java]: Please don't change the license headers.

File: src/test/java/io/reactivex/internal/subscribers/BoundedSubscriberTest.java [hunk 14]
```diff
@@ -0,0 +1,388 @@
...
+import io.reactivex.functions.Action;
+import io.reactivex.functions.Consumer;
+import io.reactivex.internal.functions.Functions;
+import io.reactivex.internal.subscriptions.BooleanSubscription;
+import io.reactivex.plugins.RxJavaPlugins;
+import io.reactivex.processors.PublishProcessor;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+import org.testng.annotations.Test;
```
Comment (akarnokd) [inline src/test/java/io/reactivex/internal/subscribers/BoundedSubscriberTest.java]: Please use `org.junit.Test` as the TestNG tests are not part of the code coverage.

**Review by akarnokd — CHANGES_REQUESTED** [review]
Fix headers and use org.junit.Test please.

**Review by akarnokd — APPROVED** [review]
